cni/pkg/taint/taintcontroller_test.go

```diff
@@ -0,0 +1,498 @@
+package taint
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	fcache "k8s.io/client-go/tools/cache/testing"
+	"k8s.io/client-go/util/workqueue"
+
+	"istio.io/istio/pkg/test/util/retry"
+)
+
+//simplified taintsetter controller build upon fake sourcer, return controller and namespace, label based sourcer created by configmap
+func newMockTaintSetterController(ts *TaintSetter, nodeSource *fcache.FakeControllerSource) (c *Controller, sourcer map[string]map[string]*fcache.FakeControllerSource, e error) {
+	c = &Controller{
+		clientset:       ts.Client,
+		podWorkQueue:    workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		nodeWorkQueue:   workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		taintsetter:     ts,
+		cachedPodsStore: make(map[string]map[string]cache.Store),
+	}
+	//construct a series of pod controller according to the configmaps' namespace and labelselector
+	c.podController = []cache.Controller{}
+	sourcer = make(map[string]map[string]*fcache.FakeControllerSource)
+	for _, config := range ts.configs {
+		podSource := fcache.NewFakeControllerSource()
+		if _, ok := sourcer[config.Namespace]; !ok {
+			sourcer[config.Namespace] = make(map[string]*fcache.FakeControllerSource)
+		}
+		sourcer[config.Namespace][config.LabelSelector] = podSource
+		tempcontroller := buildMockPodController(c, config, podSource)
+		c.podController = append(c.podController, tempcontroller)
+	}
+	c.nodeStore, c.nodeController = buildNodeControler(c, nodeSource)
+	return c, sourcer, nil
+}
+
+//mocking api for podsController
+func buildMockPodController(c *Controller, config ConfigSettings, source cache.ListerWatcher) cache.Controller {
```

Comment:
This duplication is a code smell. Why can we not reuse the main code? Aside from code duplication I am concerned this will not test the real code that is running and miss bugs